src/profile-logic/profile-data.js

```diff
@@ -150,6 +150,186 @@ export function getSampleCallNodes(
+    value: Milliseconds,
+    breakdownByImplementation: Breakdown,
+  |},
+  totalTime: {|
+    // time spent including children
+    value: Milliseconds,
+    breakdownByImplementation: Breakdown,
+  |},
+|};
+
+export type TimingsForPath = {|
+  // timings for this path
+  forPath: ItemTimings,
+  // timings for this func across the tree
+  forFunc: ItemTimings,
+  rootTime: Milliseconds, // time for all the samples in the current tree
+|};
+
+// Returns the timings for a specific path. The algorithm is adjusted when
+// the call tree is inverted.
+export function getTimingsForPath(
+  path: CallNodePath,
+  { callNodeTable, stackIndexToCallNodeIndex }: CallNodeInfo,
+  interval: number,
+  isInvertedTree: boolean,
+  thread: Thread
+): TimingsForPath {
+  const { samples, stackTable, funcTable } = thread;
+  const nodeIndex = getCallNodeIndexFromPath(path, callNodeTable);
+  const funcIndex = getFuncIndex(path);
+
+  const pathTimings: ItemTimings = {
+    selfTime: { value: 0, breakdownByImplementation: null },
+    totalTime: { value: 0, breakdownByImplementation: null },
+  };
+  const funcTimings: ItemTimings = {
+    selfTime: { value: 0, breakdownByImplementation: null },
+    totalTime: { value: 0, breakdownByImplementation: null },
+  };
+  let rootTime = 0;
+
+  // Utility function to more easily add data to breakdowns
+  function accumulateDataToBreakdown(
+    timings: { breakdownByImplementation: Breakdown },
+    implementation: string
+  ): void {
+    if (timings.breakdownByImplementation === null) {
+      timings.breakdownByImplementation = {};
+    }
+    if (timings.breakdownByImplementation[implementation] === undefined) {
+      timings.breakdownByImplementation[implementation] = 0;
+    }
+    timings.breakdownByImplementation[implementation] += interval;
+  }
+
+  // Looping over the samples' stacks
+  for (const stackIndex of samples.stack) {
+    if (stackIndex === null) {
+      continue;
+    }
+
+    rootTime += interval;
+
+    const thisNodeIndex = stackIndexToCallNodeIndex[stackIndex];
+    const thisFunc = callNodeTable.func[thisNodeIndex];
+
+    // For non-inverted trees, we compute the self time from the stacks' leaf nodes.
+    if (!isInvertedTree) {
+      if (thisNodeIndex === nodeIndex) {
+        pathTimings.selfTime.value += interval;
+
+        let implementation = 'native';
+        if (funcTable.isJS[thisFunc]) {
+          implementation = getJsImplementationForStack(stackIndex, thread);
+        }
+        accumulateDataToBreakdown(pathTimings.selfTime, implementation);
+      }
+
+      if (thisFunc === funcIndex) {
+        funcTimings.selfTime.value += interval;
+      }
+    }
+
+    // Use the stackTable to traverse the call node path and get various
+    // measurements.
+    // We don't use getCallNodePathFromIndex because we don't need the result
+    // itself, and it's costly to get.
+    let funcFound = false;
+    let pathFound = false;
+    let nextIndex;
+    for (
```

Comment:
My suspicion is that we could cache some of this information with `Map`. I'm having trouble reasoning completely with it right now though.

Reply:
From my tests on the very big profile `280e10708c0660695ef5af145c53179d29d9796c`, this seems fast enough like this already, by looping the samples only once, and that's why I didn't try to optimize more. Granted my computer is quite fast.

But yeah otherwise a cache (map or array) could help if we see issues later, especially for inverted trees where we have to loop to the root for all samples. I'll look again anyway to make sure.

Reply:
I tried on the inverted tree for the profile mentioned above, and I _do_ see a slowdown with the patch. Is it OK that I optimize this more in a subsequent PR? (possibly with a cache as you suggest)

Reply:
Yeah that'll work.